#### Задача:
Применим линейную регрессию – попробуем предсказать стоимость машин и понять, от каких факторов зависит ценообразование на автомобили. Помимо этого узнаем, какие переменные важны для прогнозирования и насколько хорошо полученная модель описывает данные.

    Регрессио́нный анализ — набор статистических методов исследования влияния одной или нескольких независимых переменных X_{1},X_{2},...,X_{p} на зависимую переменную Y.  Терминология зависимых и независимых переменных отражает лишь математическую зависимость переменных (Корреляция), а не причинно-следственные отношения.
     
      Наиболее распространённый вид регрессионного анализа — линейная регрессия, когда находят линейную функцию, которая, согласно определённым математическим критериям, наиболее соответствует данным. Например, в методе наименьших квадратов вычисляется прямая(или гиперплоскость), сумма квадратов между которой и данными минимальна.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

%matplotlib inline 

In [2]:
from scipy import stats
import statsmodels.api as sm


In [3]:
cars = pd.read_csv('https://stepik.org/media/attachments/lesson/387691/cars.csv')

In [4]:
cars.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

In [5]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

In [6]:
cars.isna().sum()

car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

Использовать полное название машины – не самый хороший вариант, поэтому создадим новый признак – марку автомобиля (company). 

In [7]:
cars['mark'] = cars.CarName.apply(lambda x: x.split(' ')[0])

In [8]:
cars.drop(columns=['car_ID','CarName'],inplace=True)

In [9]:
cars.mark.unique()

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'maxda', 'mazda', 'buick', 'mercury',
       'mitsubishi', 'Nissan', 'nissan', 'peugeot', 'plymouth', 'porsche',
       'porcshce', 'renault', 'saab', 'subaru', 'toyota', 'toyouta',
       'vokswagen', 'volkswagen', 'vw', 'volvo'], dtype=object)

In [10]:
cars.mark.replace({'maxda':'mazda',
                  'Nissan':'nissan',
                  'porcshce':'porsche',
                  'toyouta':'toyota',
                  'vokswagen':'volkswagen',
                  'vw':'volkswagen',
                  'alfa-romero':'alfa-romeo'}, inplace=True)

In [11]:
cars.mark.nunique()

22

In [12]:
cars

symboling fueltype aspiration doornumber      carbody drivewheel  \
0            3      gas        std        two  convertible        rwd   
1            3      gas        std        two  convertible        rwd   
2            1      gas        std        two    hatchback        rwd   
3            2      gas        std       four        sedan        fwd   
4            2      gas        std       four        sedan        4wd   
..         ...      ...        ...        ...          ...        ...   
200         -1      gas        std       four        sedan        rwd   
201         -1      gas      turbo       four        sedan        rwd   
202         -1      gas        std       four        sedan        rwd   
203         -1   diesel      turbo       four        sedan        rwd   
204         -1      gas      turbo       four        sedan        rwd   

    enginelocation  wheelbase  carlength  carwidth  ...  fuelsystem  \
0            front       88.6      168.8      64.1  ...        mpfi   
1            front       88.6      168.8      64.1  ...        mpfi   
2            front       94.5      171.2      65.5  ...        mpfi   
3            front       99.8      176.6      66.2  ...        mpfi   
4            front       99.4      176.6      66.4  ...        mpfi   
..             ...        ...        ...       ...  ...         ...   
200          front      109.1      188.8      68.9  ...        mpfi   
201          front      109.1      188.8      68.8  ...        mpfi   
202          front      109.1      188.8      68.9  ...        mpfi   
203          front      109.1      188.8      68.9  ...         idi   
204          front      109.1      188.8      68.9  ...        mpfi   

     boreratio stroke compressionratio  horsepower peakrpm  citympg  \
0         3.47   2.68              9.0         111    5000       21   
1         3.47   2.68              9.0         111    5000       21   
2         2.68   3.47              9.0         154    5000       19   
3         3.19   3.40             10.0         102    5500       24   
4         3.19   3.40              8.0         115    5500       18   
..         ...    ...              ...         ...     ...      ...   
200       3.78   3.15              9.5         114    5400       23   
201       3.78   3.15              8.7         160    5300       19   
202       3.58   2.87              8.8         134    5500       18   
203       3.01   3.40             23.0         106    4800       26   
204       3.78   3.15              9.5         114    5400       19   

     highwaympg    price        mark  
0            27  13495.0  alfa-romeo  
1            27  16500.0  alfa-romeo  
2            26  16500.0  alfa-romeo  
3            30  13950.0        audi  
4            22  17450.0        audi  
..          ...      ...         ...  
200          28  16845.0       volvo  
201          25  19045.0       volvo  
202          23  21485.0       volvo  
203          27  22470.0       volvo  
204          25  22625.0       volvo  

[205 rows x 25 columns]

Чтобы не перегружать модель большим количеством предикторов(yезависимые переменные), оставим только часть из них.

In [13]:
cars = cars[['mark', 'fueltype', 'aspiration','carbody', 'drivewheel', 'wheelbase', 'carlength',
'carwidth', 'curbweight', 'enginetype', 'cylindernumber', 'enginesize', 'boreratio','horsepower','price']]


После этого посчитаем **корреляцию** между price и другими переменными. Чему равна корреляция между, например, price и horsepower?

In [14]:
cars.corr().round(2)

wheelbase  carlength  carwidth  curbweight  enginesize  boreratio  \
wheelbase        1.00       0.87      0.80        0.78        0.57       0.49   
carlength        0.87       1.00      0.84        0.88        0.68       0.61   
carwidth         0.80       0.84      1.00        0.87        0.74       0.56   
curbweight       0.78       0.88      0.87        1.00        0.85       0.65   
enginesize       0.57       0.68      0.74        0.85        1.00       0.58   
boreratio        0.49       0.61      0.56        0.65        0.58       1.00   
horsepower       0.35       0.55      0.64        0.75        0.81       0.57   
price            0.58       0.68      0.76        0.84        0.87       0.55   

            horsepower  price  
wheelbase         0.35   0.58  
carlength         0.55   0.68  
carwidth          0.64   0.76  
curbweight        0.75   0.84  
enginesize        0.81   0.87  
boreratio         0.57   0.55  
horsepower        1.00   0.81  
price             0.81   1.00

In [15]:
cars.dtypes

mark               object
fueltype           object
aspiration         object
carbody            object
drivewheel         object
wheelbase         float64
carlength         float64
carwidth          float64
curbweight          int64
enginetype         object
cylindernumber     object
enginesize          int64
boreratio         float64
horsepower          int64
price             float64
dtype: object

Последний шаг в подготовке данных: линейная регрессия в python не справляется с категориальными переменными (типом object в pandas), поэтому давайте применим **pd.get_dummies()** (из категориальных в индикаторные переменные)

In [16]:
df_dummy = pd.get_dummies(data=cars[['mark','fueltype','aspiration','carbody','drivewheel','enginetype','cylindernumber']],\
                          drop_first = True)

In [17]:
df_dummy

mark_audi  mark_bmw  mark_buick  mark_chevrolet  mark_dodge  mark_honda  \
0            0         0           0               0           0           0   
1            0         0           0               0           0           0   
2            0         0           0               0           0           0   
3            1         0           0               0           0           0   
4            1         0           0               0           0           0   
..         ...       ...         ...             ...         ...         ...   
200          0         0           0               0           0           0   
201          0         0           0               0           0           0   
202          0         0           0               0           0           0   
203          0         0           0               0           0           0   
204          0         0           0               0           0           0   

     mark_isuzu  mark_jaguar  mark_mazda  mark_mercury  ...  enginetype_ohc  \
0             0            0           0             0  ...               0   
1             0            0           0             0  ...               0   
2             0            0           0             0  ...               0   
3             0            0           0             0  ...               1   
4             0            0           0             0  ...               1   
..          ...          ...         ...           ...  ...             ...   
200           0            0           0             0  ...               1   
201           0            0           0             0  ...               1   
202           0            0           0             0  ...               0   
203           0            0           0             0  ...               1   
204           0            0           0             0  ...               1   

     enginetype_ohcf  enginetype_ohcv  enginetype_rotor  cylindernumber_five  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                1                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    1   
..               ...              ...               ...                  ...   
200                0                0                 0                    0   
201                0                0                 0                    0   
202                0                1                 0                    0   
203                0                0                 0                    0   
204                0                0                 0                    0   

     cylindernumber_four  cylindernumber_six  cylindernumber_three  \
0                      1                   0                     0   
1                      1                   0                     0   
2                      0                   1                     0   
3                      1                   0                     0   
4                      0                   0                     0   
..                   ...                 ...                   ...   
200                    1                   0                     0   
201                    1                   0                     0   
202                    0                   1                     0   
203                    0                   1                     0   
204                    1                   0                     0   

     cylindernumber_twelve  cylindernumber_two  
0                        0                   0  
1                        0                   0  
2                        0                   0  
3                        0                   0  
4                        0                   0  

In [18]:
cars.drop(columns=['mark','fueltype','aspiration','carbody','drivewheel','enginetype','cylindernumber'], inplace=True)

C:\Users\maksa\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [19]:
cars

wheelbase  carlength  carwidth  curbweight  enginesize  boreratio  \
0         88.6      168.8      64.1        2548         130       3.47   
1         88.6      168.8      64.1        2548         130       3.47   
2         94.5      171.2      65.5        2823         152       2.68   
3         99.8      176.6      66.2        2337         109       3.19   
4         99.4      176.6      66.4        2824         136       3.19   
..         ...        ...       ...         ...         ...        ...   
200      109.1      188.8      68.9        2952         141       3.78   
201      109.1      188.8      68.8        3049         141       3.78   
202      109.1      188.8      68.9        3012         173       3.58   
203      109.1      188.8      68.9        3217         145       3.01   
204      109.1      188.8      68.9        3062         141       3.78   

     horsepower    price  
0           111  13495.0  
1           111  16500.0  
2           154  16500.0  
3           102  13950.0  
4           115  17450.0  
..          ...      ...  
200         114  16845.0  
201         160  19045.0  
202         134  21485.0  
203         106  22470.0  
204         114  22625.0  

[205 rows x 8 columns]

In [20]:
# присоединим к полученному датасету столбцы с переменными других типов
cars_ind = pd.concat([cars,df_dummy], axis=1)

In [21]:
cars_ind

wheelbase  carlength  carwidth  curbweight  enginesize  boreratio  \
0         88.6      168.8      64.1        2548         130       3.47   
1         88.6      168.8      64.1        2548         130       3.47   
2         94.5      171.2      65.5        2823         152       2.68   
3         99.8      176.6      66.2        2337         109       3.19   
4         99.4      176.6      66.4        2824         136       3.19   
..         ...        ...       ...         ...         ...        ...   
200      109.1      188.8      68.9        2952         141       3.78   
201      109.1      188.8      68.8        3049         141       3.78   
202      109.1      188.8      68.9        3012         173       3.58   
203      109.1      188.8      68.9        3217         145       3.01   
204      109.1      188.8      68.9        3062         141       3.78   

     horsepower    price  mark_audi  mark_bmw  ...  enginetype_ohc  \
0           111  13495.0          0         0  ...               0   
1           111  16500.0          0         0  ...               0   
2           154  16500.0          0         0  ...               0   
3           102  13950.0          1         0  ...               1   
4           115  17450.0          1         0  ...               1   
..          ...      ...        ...       ...  ...             ...   
200         114  16845.0          0         0  ...               1   
201         160  19045.0          0         0  ...               1   
202         134  21485.0          0         0  ...               0   
203         106  22470.0          0         0  ...               1   
204         114  22625.0          0         0  ...               1   

     enginetype_ohcf  enginetype_ohcv  enginetype_rotor  cylindernumber_five  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                1                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    1   
..               ...              ...               ...                  ...   
200                0                0                 0                    0   
201                0                0                 0                    0   
202                0                1                 0                    0   
203                0                0                 0                    0   
204                0                0                 0                    0   

     cylindernumber_four  cylindernumber_six  cylindernumber_three  \
0                      1                   0                     0   
1                      1                   0                     0   
2                      0                   1                     0   
3                      1                   0                     0   
4                      0                   0                     0   
..                   ...                 ...                   ...   
200                    1                   0                     0   
201                    1                   0                     0   
202                    0                   1                     0   
203                    0                   1                     0   
204                    1                   0                     0   

     cylindernumber_twelve  cylindernumber_two  
0                        0                   0  
1                        0                   0  
2                        0                   0  
3                        0                   0  
4                        0                   0  
..                     ...                 ...  
200                      0                   0  
201                      0                   0  
202                      0      

Сначала построим небольшую модель всего с одним предиктором цены (price) – horsepower.

Какой процент изменчивости объясняет полученная модель?

In [22]:
cars_ind.shape

(205, 49)

In [23]:
import statsmodels.formula.api as smf

In [24]:
res = smf.ols('price~horsepower', data=cars_ind).fit()
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.653
Model:                            OLS   Adj. R-squared:                  0.651
Method:                 Least Squares   F-statistic:                     382.2
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           1.48e-48
Time:                        22:48:28   Log-Likelihood:                -2024.0
No. Observations:                 205   AIC:                             4052.
Df Residuals:                     203   BIC:                             4059.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept  -3721.7615    929.849     -4.003      0.000   -5555.163   -1888.360
horsepower   163.2631      8.351     19.549      0.000     146.796     179.730
==============================================================================
Omnibus:                       47.741   Durbin-Watson:                   0.792
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               91.702
Skew:                           1.141   Prob(JB):                     1.22e-20
Kurtosis:                       5.352   Cond. No.                         314.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

процент изменчивости объясняет полученная модель: **R-squared	0.653**

Теперь – две модели:

    модель со всеми предикторами
    модель со всеми предикторами, кроме марок машин

In [25]:
import statsmodels.api as sm

In [26]:
Y =  cars_ind['price']
x1 = cars_ind.drop(columns=['price'])

In [27]:
x1 = sm.add_constant(x1)  # добавить константу, чтобы был свободный член
model = sm.OLS(Y, x1)  # говорим модели, что у нас ЗП, а что НП
results = model.fit()  # строим регрессионную прямую
print(results.summary())  # смотрим результат


                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.959
Model:                            OLS   Adj. R-squared:                  0.948
Method:                 Least Squares   F-statistic:                     81.09
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           4.86e-89
Time:                        22:48:28   Log-Likelihood:                -1804.2
No. Observations:                 205   AIC:                             3702.
Df Residuals:                     158   BIC:                             3858.
Df Model:                          46                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                 -3.472e+

In [28]:
cars_ind.columns

Index(['wheelbase', 'carlength', 'carwidth', 'curbweight', 'enginesize',
       'boreratio', 'horsepower', 'price', 'mark_audi', 'mark_bmw',
       'mark_buick', 'mark_chevrolet', 'mark_dodge', 'mark_honda',
       'mark_isuzu', 'mark_jaguar', 'mark_mazda', 'mark_mercury',
       'mark_mitsubishi', 'mark_nissan', 'mark_peugeot', 'mark_plymouth',
       'mark_porsche', 'mark_renault', 'mark_saab', 'mark_subaru',
       'mark_toyota', 'mark_volkswagen', 'mark_volvo', 'fueltype_gas',
       'aspiration_turbo', 'carbody_hardtop', 'carbody_hatchback',
       'carbody_sedan', 'carbody_wagon', 'drivewheel_fwd', 'drivewheel_rwd',
       'enginetype_dohcv', 'enginetype_l', 'enginetype_ohc', 'enginetype_ohcf',
       'enginetype_ohcv', 'enginetype_rotor', 'cylindernumber_five',
       'cylindernumber_four', 'cylindernumber_six', 'cylindernumber_three',
       'cylindernumber_twelve', 'cylindernumber_two'],
      dtype='object')

In [29]:
x2 = cars_ind.drop(columns=['price', 'mark_audi','mark_bmw',
                            'mark_buick','mark_chevrolet', 
                            'mark_dodge','mark_honda','mark_isuzu', 
                            'mark_jaguar', 'mark_mazda', 'mark_mercury',
                            'mark_mitsubishi','mark_nissan', 'mark_peugeot',
                            'mark_plymouth','mark_porsche', 'mark_renault',
                            'mark_saab', 'mark_subaru','mark_toyota',
                            'mark_volkswagen', 'mark_volvo'])

In [30]:
x2

wheelbase  carlength  carwidth  curbweight  enginesize  boreratio  \
0         88.6      168.8      64.1        2548         130       3.47   
1         88.6      168.8      64.1        2548         130       3.47   
2         94.5      171.2      65.5        2823         152       2.68   
3         99.8      176.6      66.2        2337         109       3.19   
4         99.4      176.6      66.4        2824         136       3.19   
..         ...        ...       ...         ...         ...        ...   
200      109.1      188.8      68.9        2952         141       3.78   
201      109.1      188.8      68.8        3049         141       3.78   
202      109.1      188.8      68.9        3012         173       3.58   
203      109.1      188.8      68.9        3217         145       3.01   
204      109.1      188.8      68.9        3062         141       3.78   

     horsepower  fueltype_gas  aspiration_turbo  carbody_hardtop  ...  \
0           111             1                 0                0  ...   
1           111             1                 0                0  ...   
2           154             1                 0                0  ...   
3           102             1                 0                0  ...   
4           115             1                 0                0  ...   
..          ...           ...               ...              ...  ...   
200         114             1                 0                0  ...   
201         160             1                 1                0  ...   
202         134             1                 0                0  ...   
203         106             0                 1                0  ...   
204         114             1                 1                0  ...   

     enginetype_ohc  enginetype_ohcf  enginetype_ohcv  enginetype_rotor  \
0                 0                0                0                 0   
1                 0                0                0                 0   
2                 0                0                1                 0   
3                 1                0                0                 0   
4                 1                0                0                 0   
..              ...              ...              ...               ...   
200               1                0                0                 0   
201               1                0                0                 0   
202               0                0                1                 0   
203               1                0                0                 0   
204               1                0                0                 0   

     cylindernumber_five  cylindernumber_four  cylindernumber_six  \
0                      0                    1                   0   
1                      0                    1                   0   
2                      0                    0                   1   
3                      0                    1                   0   
4                      1                    0                   0   
..                   ...                  ...                 ...   
200                    0                    1                   0   
201                    0                    1                   0   
202                    0                    0                   1   
203                    0                    0                   1   
204                    0                    1                   0   

     cylindernumber_three  cylindernumber_twelve  cylindernumber_two  
0                       0                      0                   0  
1                       0                      0                   0  
2                       0                      0                   0  
3                       0                      0                   0  
4                       0                      0                   0  
..                    ...                    ...                 ...  
200         

In [31]:
x2 = sm.add_constant(x2)  # добавить константу, чтобы был свободный член
model = sm.OLS(Y, x2)  # говорим модели, что у нас ЗП, а что НП
results = model.fit()  # строим регрессионную прямую
print(results.summary())  # смотрим результат


                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.914
Model:                            OLS   Adj. R-squared:                  0.901
Method:                 Least Squares   F-statistic:                     72.32
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           9.86e-81
Time:                        22:48:28   Log-Likelihood:                -1881.6
No. Observations:                 205   AIC:                             3817.
Df Residuals:                     178   BIC:                             3907.
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
const                   -1.7e+

Выбранная модель объясняет примерно 
90
 % дисперсии (Adj. R-squared). 
 
 Среди предикторов 
10
 из 27 оказались не значимыми (p > 0.05). 
 
 Пример интерпретации: при единичном изменении показателя horsepower, цена 
ВОЗРАСТАЕТ
 на 
86.8164
 (без округления).

Обратите внимание на изменения в R^2, коэффициентах и их значимости. Какую модель лучше оставить? ответ: 
     
     Где меньше предикторов, ведь R^2 изменился не очень сильно, а часть марок вообще не значима